### PR TITLE
Sync node status from node controller to master.

### PR DESCRIPTION
**Pull request by deyuanix**
@bgrant0607 @lavalamp @smarterclayton @erictune @brendandburns 

e2e passes, I may pick up this issue #3520 if I can set up some time

### CODE REVIEW

File: cmd/kube-controller-manager/controller-manager.go [hunk 1]
```diff
@@ -102,6 +106,10 @@ func main() {
 	controllerManager := replicationControllerPkg.NewReplicationManager(kubeClient)
 	controllerManager.Run(10 * time.Second)
 
+	kubeletClient, err := client.NewKubeletClient(&kubeletConfig)
```
Comment (bgrant0607) [inline cmd/kube-controller-manager/controller-manager.go]: The apiserver does:
client.BindKubeletClientConfigFlags(flag.CommandLine, &kubeletConfig)

Is that done somewhere here?

File: cmd/kube-controller-manager/controller-manager.go [hunk 2]
```diff
@@ -109,8 +117,8 @@ func main() {
 			api.ResourceMemory: *nodeMemory,
 		},
 	}
-	nodeController := nodeControllerPkg.NewNodeController(cloud, *minionRegexp, machineList, nodeResources, kubeClient)
-	nodeController.Run(*nodeSyncPeriod)
+	nodeController := nodeControllerPkg.NewNodeController(cloud, *minionRegexp, machineList, nodeResources, kubeClient, kubeletClient)
+	nodeController.Run(*nodeSyncPeriod, 10)
```
Comment (bgrant0607) [inline cmd/kube-controller-manager/controller-manager.go]: What is this magic number 10? Please create a constant or something.

File: pkg/api/validation/validation.go [hunk 3]
```diff
@@ -612,9 +612,10 @@ func ValidateMinion(minion *api.Node) errs.ValidationErrorList {
 func ValidateMinionUpdate(oldMinion *api.Node, minion *api.Node) errs.ValidationErrorList {
 	allErrs := errs.ValidationErrorList{}
 
-	if !api.Semantic.DeepEqual(minion.Status, api.NodeStatus{}) {
-		allErrs = append(allErrs, errs.NewFieldInvalid("status", minion.Status, "status must be empty"))
-	}
+	// TODO: enable the code once we have better api object.status update model
```
Comment (bgrant0607) [inline pkg/api/validation/validation.go]: This makes it possible for anyone to write the node status?  Please add a comment to that effect, if true.

File: pkg/cloudprovider/controller/nodecontroller.go [hunk 4]
```diff
@@ -34,80 +35,94 @@ type NodeController struct {
...
+				return nil
 			}
 		}
-		if registered.Len() == len(s.nodes) {
-			return nil
-		}
-		time.Sleep(period)
+		time.Sleep(retryInterval)
 	}
+	return nil
```
Comment (bgrant0607) [inline pkg/cloudprovider/controller/nodecontroller.go]: This doesn't raise any kind of error if not all nodes could be registered?

File: pkg/cloudprovider/controller/nodecontroller.go [hunk 5]
```diff
@@ -120,10 +135,11 @@ func (s *NodeController) SyncCloud() error {
 		nodeMap[node.Name] = &node
 	}
 
-	// Create or delete nodes from registry.
+	// Create nodes which have been created in cloud, but not in kubernetes cluster.
 	for _, node := range matches.Items {
 		if _, ok := nodeMap[node.Name]; !ok {
 			glog.Infof("Create node in registry: %s", node.Name)
+			node.Status.Phase = api.NodePending
```
Comment (bgrant0607) [inline pkg/cloudprovider/controller/nodecontroller.go]: Do you also set the Pending phase prior to RegisterNodes?

**Comment by bgrant0607** [comment]
FYI: #3695

/cc @davidopp

File: pkg/cloudprovider/controller/nodecontroller.go [hunk 6]
```diff
@@ -139,11 +156,91 @@ func (s *NodeController) SyncCloud() error {
...
+			Kind:   api.NodeReady,
+			Status: api.ConditionNone,
+		})
+	default:
+		conditions = append(conditions, api.NodeCondition{
+			Kind:   api.NodeReady,
+			Status: api.ConditionFull,
+		})
+	}
+	glog.V(2).Infof("NodeController: node %q status was %+v", node.Name, conditions)
```
Comment (bgrant0607) [inline pkg/cloudprovider/controller/nodecontroller.go]: This should be V(5) or higher. It's too spammy to leave enabled by default.

File: pkg/cloudprovider/controller/nodecontroller.go [hunk 7]
```diff
@@ -139,11 +156,91 @@ func (s *NodeController) SyncCloud() error {
...
+			glog.Errorf("error updating node %s: %v", node.Name, err)
+		}
+	}
+	return nil
+}
+
+// DoChecks performs health checking for given list of nodes.
+func (s *NodeController) DoChecks(nodes *api.NodeList) *api.NodeList {
+	for i := range nodes.Items {
+		// TODO: Add concurrency.
```
Comment (bgrant0607) [inline pkg/cloudprovider/controller/nodecontroller.go]: Could we just add concurrency now?

http://golang.org/pkg/sync/#WaitGroup could be used to wait for completion before returning.

BTW, what happens if the checks take longer than the invocation period?

**Comment by bgrant0607** [comment]
Cool. Happy to see this. Thanks.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 7 above)
Comment (ddysher) [inline pkg/cloudprovider/controller/nodecontroller.go]: Oh, no... I've already cleared the TODO.  I rebased the PR on another machine, and use push -f to update the PR.  I'll check the other machine, that one probably has the latest changes.

File: cmd/kube-controller-manager/controller-manager.go (hunk 1 above)
Comment (ddysher) [inline cmd/kube-controller-manager/controller-manager.go]: nop, updated.

File: cmd/kube-controller-manager/controller-manager.go (hunk 2 above)
Comment (ddysher) [inline cmd/kube-controller-manager/controller-manager.go]: Hmm...well, we are using magic number quite a frew places, so I haven't pay much attention to it.  Changed to a flag.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 4 above)
Comment (ddysher) [inline pkg/cloudprovider/controller/nodecontroller.go]: We should.  This is left from previous code, updated.

File: pkg/api/validation/validation.go (hunk 3 above)
Comment (ddysher) [inline pkg/api/validation/validation.go]: Yes.  Added comment.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 5 above)
Comment (ddysher) [inline pkg/cloudprovider/controller/nodecontroller.go]: In the newer version, I removed node phase.

I'd like to have another PR handling phase update.  Currently all k8s components do not respect node phase, so I think we are good.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 7 above)
Comment (ddysher) [inline pkg/cloudprovider/controller/nodecontroller.go]: Regarding to checks taking longer than invocation period.  I think we are fine cause util.Forever() invokes checks every period time, the time does not include the method running time.

**Comment by ddysher** [comment]
As mentioned in one of the inline comment, I accidentally pushed an older working commit to the PR. However, much of the structure are the same.

Comment addressed.  PTAL.

**Comment by bgrant0607** [comment]
LGTM.
Heads up @mikedanese

**Comment by bgrant0607** [comment]
Rerunning travis.

**Comment by bgrant0607** [comment]
Waiting on #3695 to go in. Sorry @ddysher, but you'll need to rebase.

**Comment by bgrant0607** [comment]
#3695 was merged.

**Comment by bgrant0607** [comment]
#3818 was merged, also. Please rebase.

**Comment by ddysher** [comment]
@bgrant0607 will rebase after work

**Comment by ddysher** [comment]
Rebased.  PTAL.  Also e2e passed with latest e2e fixes.

**Comment by bgrant0607** [comment]
Thanks. LGTM.

**Comment by bgrant0607** [comment]
Waiting for another travis run before merging.

**Comment by bgrant0607** [comment]
Sorry, @ddysher. Needs another rebase, apparently.

**Comment by ddysher** [comment]
ok, resolved two conflicts 
FYI:
First one is rest_test.go for minion registry, I've deleleted some tests related to health check, but PR#3856 touches those.
Second is auth_test.go, the url param change in PR#3707 has emerged the 409 error code (#2115).  Which is the right error code since we allow minion update from thirdparty for not.

**Comment by bgrant0607** [comment]
Thanks, @ddysher. Yes, 409 is the right error code for updates.

LGTM. Will first first thing in the morning.

**Comment by davidopp** [comment]
Can we merge this now (after a rebase)?

**Comment by bgrant0607** [comment]
This was merged days ago. Wrong PR?

**Pull request merged**